src/kernel.ts

```diff
@@ -3,1067 +3,792 @@
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
 
-    _kernel_connected(): void {
-        /**
-         * Perform necessary tasks once the connection to the kernel has
-         * been established. This includes requesting information about
-         * the kernel.
-         *
-         * @function _kernel_connected
-         */
-        events.trigger('kernel_connected.Kernel', {kernel: this});
-        // get kernel info so we know what state the kernel is in
-        var that = this;
-        this.kernel_info((reply?) => {
-            this.info_reply = reply.content;
-            events.trigger('kernel_ready.Kernel', {kernel: that});
-        });
-    }
+/**
+ * Implementation of a Kernel Future.
+ */
+class KernelFutureHandler extends Disposable implements IKernelFuture {
+
+  /**
+   * Dispose and unregister the future.
+   */
+  dispose(): void {
+    super.dispose();
+    this._input = null;
+    this._output = null;
+    this._reply = null;
+    this._done = null;
+  }
+
+  /**
+   * Register a reply handler. Returns 'this'.
+   */
+  onReply(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._reply = cb;
+    return this;
+  }
+
+  /**
+   * Register an output handler. Returns 'this'.
+   */
+  onOutput(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._output = cb;
+    return this;
+  }
+
+  /**
+   * Register a done handler. Returns 'this'.
+   */
+  onDone(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._done = cb;
+    return this;
+  }
+
+  /**
+   * Register an input handler. Returns 'this'.
+   */
+  onInput(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._input = cb;
+    return this;
+  }
+
+  /**
+   * Get the current autoDispose status.
+   */
+  get autoDispose(): boolean {
+    return this._testFlag(KernelFutureFlag.AutoDispose);
+  }
+
+  /**
+   * Set the current autoDispose behavior of the Feature.
+   *
+   * If True, it will self-dispose() after onDone() is called.
+   */
+  set autoDispose(value: boolean) {
+    if (value) {
+      this._setFlag(KernelFutureFlag.AutoDispose);
+    } else {
+      this._clearFlag(KernelFutureFlag.AutoDispose);
+    }
+  }
+
+  /**
+   * Handle an incoming message from the kernel belonging to this Future.
+   */
+  handleMsg(msg: IKernelMsg): void {
+    if (msg.channel === 'iopub') {
+      if (msg.msgType === 'status' && msg.content.execution_state === 'idle') {
+        this._setFlag(KernelFutureFlag.GotIdle);
+        if (this._testFlag(KernelFutureFlag.GotReply)) {
+          this._handleDone(msg);
+        }
+      }
+      if (this._output) {
+        this._output(msg);
+      }
+    } else if (msg.channel === 'shell') {
+      this._setFlag(KernelFutureFlag.GotReply)
+      if (this._testFlag(KernelFutureFlag.GotIdle)) {
+        this._handleDone(msg);
+      }
+      if (this._reply) {
```

Comment:
I think the reply handler will never be called, because if you call handleDone above, it deletes the reply handler.  We have the same issue with the output message above, but that probably won't matter since the idle message should come after any output message (but if we keep the code as-is, we should explicitly say that the output handler will not be called with the last iopub status idle message).